### ISSUE TITLE
Deadlock When Running distributed.initialize() on one TPU Host by Accident

### ISSUE THREAD

**Issue by s-smits** [google/jax#22436]
## Issue
Encountered a deadlock while running a JAX-based LLM training script on a TPU-v4-32 pod. SSH'd into worker 0 and ran the script there directly, instead of using `--worker all --command "..."`.

## Command

```bash
python -m EasyLM.models.llama.llama_train \
    --mesh_dim='-1,4,1' \
    --dtype='fp16' \
    --total_steps=250000 \
    --log_freq=50 \
    --save_model_freq=0 \
    --save_milestone_freq=2500 \
    --load_llama_config='7b' \
    --update_llama_config='' \
    --load_dataset_state='' \
    --load_checkpoint='' \
    --optimizer.type='adamw' \
    --optimizer.adamw_optimizer.weight_decay=0.1 \
    --optimizer.adamw_optimizer.lr=3e-4 \
    --optimizer.adamw_optimizer.end_lr=3e-5 \
    --optimizer.adamw_optimizer.lr_warmup_steps=2000 \
    --optimizer.adamw_optimizer.lr_decay_steps=250000 \
    --train_dataset.type='json' \
    --train_dataset.text_processor.fields='text' \
    --train_dataset.json_dataset.path='togethercomputer/RedPajama-Data-1T' \
    --train_dataset.json_dataset.seq_length=2048 \
    --train_dataset.json_dataset.batch_size=2048 \
    --train_dataset.json_dataset.tokenizer_processes=16 \
    --checkpointer.save_optimizer_state=True \
    --logger.online=True \
    --logger.prefix='EasyLM' \
    --logger.project="open_llama_7b" \
    --logger.output_dir="." \
    --logger.wandb_dir="$HOME/experiment_output/open_llama_7b"
```

## Error Message
```
/home/air/.local/lib/python3.10/site-packages/jax/_src/xla_bridge.py:152: UserWarning: TPU backend initialization is taking more than 60.0 seconds. Did you run your code on all TPU hosts? See https://jax.readthedocs.io/en/latest/multi_process.html for more information.
  warnings.warn(
```

After this warning, the script hung indefinitely.

## Questions
1. Is this a known issue when running multi-host JAX code on a single host?
2. How should the command be modified to run correctly on a single TPU host?
3. Are there recommended checks to prevent this type of deadlock?
4. Is there a way to handle this situation gracefully in the code?

## Environment
- JAX version: 0.4.28
- Hardware: TPU-v4-32 pod
- Python version: 3.10
- TPU-Ubuntu-base-22.04

**Comment by skye** [comment]
@s-smits what do you think would be the best behavior here? We could eventually time out completely and exit with an error instead of hanging, but the timeout would have to be pretty long to make sure we don't accidentally quit too early on large deployments. I'm curious if you have more ideas!

**Comment by s-smits** [comment]
Yes, that sounds very tricky indeed. A time-out is not needed in my opinion. Being able to CTRL+C to cancel the TPU processes initiated with distributed.initialize() would solve the problem for me. I'd reckon the nesting depth is quite large, with multiple layers of TPU processes, such that even repeated CTRL+C attempts fail to fully terminate the entire process tree, however this would be desirable. If this is an unique edge case and you've never seen such a request before, I don't mind closing the Issue without a solution.

**Comment by skye** [comment]
Ah great point about CTRL+C not working! I've also run into this, although haven't investigated yet. I think it does have something to do with calling into the C++ TPU runtime.

Let's keep this issue open to track handling CTRL+C correctly in this situation. I took the liberty of editing your initial issue a bit to include this, so people don't have to read the whole thread (feel free to edit my edits more). Thanks for this feedback!
